chain/network/src/peer_manager/peer_manager_actor.rs

```diff
@@ -125,6 +127,58 @@ pub(crate) struct ConnectedPeer {
     encoding: Option<Encoding>,
 }
 
+#[derive(Default)]
+pub(crate) struct ConnectedPeers(ArcSwap<im::HashMap<PeerId, ConnectedPeer>>);
+
+impl ConnectedPeers {
+    fn read(&self) -> Arc<im::HashMap<PeerId, ConnectedPeer>> {
+        self.0.load_full()
+    }
+
+    fn update(&self, mut f: impl FnMut(&mut im::HashMap<PeerId, ConnectedPeer>)) {
+        self.0.rcu(|peers| {
+            let mut peers: im::HashMap<PeerId, ConnectedPeer> = (**peers).clone();
+            f(&mut peers);
+            Arc::new(peers)
+        });
+    }
+
+    fn insert(&self, peer: ConnectedPeer) {
+        self.update(|peers| {
+            peers.insert(peer.full_peer_info.peer_info.id.clone(), peer.clone());
+        });
+    }
+
+    fn remove(&self, peer_id: &PeerId) {
+        self.update(|peers| {
+            peers.remove(peer_id);
+        });
+    }
+
+    fn set_last_time_peer_requested(&self, peer_id: &PeerId, t: time::Instant) {
+        self.update(|peers| {
+            peers.get_mut(peer_id).map(|p| p.last_time_peer_requested = t);
```

Comment:
not a huge fan of side-effectful maps, I'd

```
if let Some(p) = peer.get_mut(peer_id) {
    p.last_time_peer_requested = t;
}
```

Reply:
fair enough. done